### PR TITLE
Macro Optimizes Map Saving (100x) DO NOT CHURN STRINGS Edition 

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Yello!
This one is reasonably quick, tho I did some fixes too

This is the big one, fixes the buildmode tool sometimes locking disabled for the whole round.
We do this by replacing the static var on buildmode with global var and a global proc
This keeps a harddel on the buildmode datum from permalocking is_running to TRUE

Also makes flipping the var BACK if something breaks significantly easier for admins, so that's nice

Alright, smaller things now

Fixes lists of numbers failing to encoded improperly This was fixed on shiptest, we failed to actually port their most recent revision
Fixes the shuttle flag not actually working because it used istype instead of ispath
Changes obj_blacklist to a typecache for optimization's sake
Renames/moves some vars around to prevent weird double typing things
Removes a checktick in key gen, it's just costing more time then it would save in overtime
Properly handles lists. We were only doing var encoding one layer deep, need to do it alll the way down

Alright, now the optimizations

This proc is fucking HOT, and it's for really dumb reasons

This is a text gen proc, and it makes the mistake of generating text and concatinating it with MORE text. 
This is HORRIFICALLY EXPENSIVE because byond caches strings (can only be one of each) and string churn fucks up that caching system something fierce
Moving from strings to lists of strings we join at the end takes us from like idk 100 seconds to save bare metastation to like 1.5
This is applied basically everywhere for obvious reasons

While I'm here, storing keys in a flat list and then using find to find them, then using that index to lookup into another flat list is a bit silly. Let's just make it an assoc list. Faster lookup, cleaner.

Oh also rather then iterating over all the vars on an object, let's iterate over just the ones we care about yeah?

Let's see... no sense genning a key we'll never use, and having suffixes be often non existent is silly just embrace the slight mess.

That's it I think, this takes us from 100 seconds to save metastation to 2.5 seconds to save ALL of metastation (I removed the vars limiter so I could make sure var saving didn't fuck me up)

## Why It's Good For The Game

Cleans up some issues that we failed to port the fixes for, MASSIVELY optimizes this (so it can finish in like 5/10 seconds and not 300!) and ensures admins can always use the thing and don't risk dropping their pet buildastation to the void.

Worth noting, this tool really should not be used for station mapping outside an event context. It produces sorta buggy var edits, and WILL fail to pull over context for shit. Please don't use it as such

Profiles (csv files I promise)
[Before](https://github.com/tgstation/tgstation/files/13853313/profiler.json)
[After](https://github.com/tgstation/tgstation/files/13853271/profiler.json)

I'd include my line by lines but I don't know how much you'd get out of them. Here's an image tho

image

## Changelog
:cl:
fix: The map saving tool will no longer lock up and prevent all further action at random
fix: Map saving now takes on the order of seconds, not minutes
fix: Fixes an issue with lists that caused strongdmm to report saved maps as broken
/:cl:

### CODE REVIEW

File: code/modules/admin/verbs/map_export.dm [hunk 1]
```diff
@@ -164,56 +188,57 @@ GLOBAL_LIST_INIT(save_file_chars, list(
...
 				//====Get turfs Data====
-				var/turf/place = locate((minx + x), (miny + y), (minz + z))
+				var/turf/place
 				var/area/location
-				var/list/objects
-				var/area/place_area = get_area(place)
+				var/turf/pull_from = locate((minx + x), (miny + y), (minz + z))
 				//If there is nothing there, save as a noop (For odd shapes)
-				if(!place)
+				if(!pull_from)
```
Comment (san7890) [inline code/modules/admin/verbs/map_export.dm:205]:
```suggestion
				if(isnull(pull_from))
```

shoulda been like this to start imo- musta missed it the first time

**Review by san7890 — CHANGES_REQUESTED** [review]
nit, we're fine otherwise

File: code/modules/admin/verbs/map_export.dm [hunk 2]
```diff
@@ -234,65 +259,50 @@ GLOBAL_LIST_INIT(save_file_chars, list(
...
 	var/list/vars_to_save = object.get_save_vars()
-	if(!vars_to_save)
-		return
-	for(var/variable in object.vars)
+	for(var/variable in vars_to_save)
 		CHECK_TICK
-		if(!(variable in vars_to_save))
-			continue
 		var/value = object.vars[variable]
 		if(!value)
```
Comment (Ghommie) [inline code/modules/admin/verbs/map_export.dm:281]: I see a small problem here. Aargs with a value of 0, "" or null won't be saved.

File: code/modules/admin/verbs/map_export.dm [hunk 3]
```diff
@@ -234,65 +259,50 @@ GLOBAL_LIST_INIT(save_file_chars, list(
...
-			symbol = "\""
-			value = hashtag_newlines_and_tabs(value, list("{"="", "}"="", "\""="", ";"="", ","=""))
-		else if(islist(value))
-			value = to_list_string(value)
-		else if(isicon(value) || isfile(value))
-			symbol = "'"
-		else if(!(isnum(value) || ispath(value)))
+
+		var/text_value = tgm_encode(value)
+		if(!text_value)
```
Comment (Ghommie) [inline code/modules/admin/verbs/map_export.dm:289]: Similar issue here, with zeros.

**Review by Ghommie — CHANGES_REQUESTED** [review]

File: code/modules/admin/verbs/map_export.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/modules/admin/verbs/map_export.dm:281]: oh right, yeah.
IDK why it's written like this, it's dumb
We really want (is the variable in vars) but that's slow as hell, maybe just write anyways idk

Comment (Ghommie) [reply code/modules/admin/verbs/map_export.dm:281]: Yeah, just write it. Negation values are valid values by `tgm_encode` standards, while real bad ones are things like live datums.

File: code/modules/admin/verbs/map_export.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/modules/admin/verbs/map_export.dm:205]: killing u/crying

File: code/modules/admin/verbs/map_export.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/modules/admin/verbs/map_export.dm:281]: I will use BOTH. I do not know if the map reader supports non existing vars and I don't really want to encode them.
this should be cheap enough to be fine

File: code/modules/admin/verbs/map_export.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/modules/admin/verbs/map_export.dm:289]: I think I will just ensure encode always returns a string

File: code/modules/admin/verbs/map_export.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/modules/admin/verbs/map_export.dm:281]: nevermind that is WAY too expensive. we'll just treat em all as valid and cry

**Pull request merged**